src/Elliptic/Actions/InitializeAnalyticSolution.hpp

```diff
@@ -41,10 +41,20 @@ namespace Actions {
  * DataBox:
  * - Adds:
  *   - `db::wrap_tags_in<::Tags::Analytic, AnalyticSolutionFields>`
+ *
+ * \note This action relies on the `SetupDataBox` aggregated initialization
+ * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
+ * phase action list prior to this action.
```

Comment:
This currently fairly "loose" documentation-based link between the initialization actions and the `SetupDataBox` mechanism seems like a prime example for a use case of a protocol :) Here's my suggestion to more strongly define and document the compile-time interface for these "initialization actions":

1. Create the following protocol in `Parallel/Protocols.hpp`:

    ```cpp
    /*!
     * \brief An action that initializes tags in the DataBox
    *
    * Actions that conform to this protocol rely on the `SetupDataBox` aggregated initialization
    * mechanism. They must provide the typelists `simple_tags` and `compute_tags`. Tags in these
    * typelists are added to the DataBox by the `SetupDataBox` action, which must be present
    * in the phase action-list prior to the initialization action. The initialization action can then
    * mutate the tags to set their initial values.
    *
    * Requires the conforming action has these type aliases:
    * - `simple_tags`: A `tmpl::list` of `db::SimpleTag`s that will be added to the DataBox and
    * initially default-constructed. The initialization action can mutate these tags to set their initial values.
    * - `compute_tags`: A `tmpl::list` of `db::ComputeTag`s that will be added to the DataBox.
    *
    * Here's an example of an action that conforms to this protocol:
    * \snippet ...
    */
    struct InitializationAction {
      template <typename ConformingType>
      struct test {
        using simple_tags = typename ConformingType::simple_tags;
        using compute_tags = typename ConformingType::compute_tags;
        // We could also assert that `simple_tags` contains no compute tags, and vice-versa
      };
    };
    ```
    Add the example snippet to `tests/Unit/Parallel/Test_Protocols.cpp` (see `tests/Unit/Evolution/Test_Protocols.cpp` for an example).
2. In `SetupDataBox`, instead of checking whether each action has `simple_tags` and `compute_tags` aliases, check `tt::conforms_to_v<Parallel::protocols::InitializationAction>`. When this is true, you can rely on the action having the aliases.
3. Derive all initialization actions from `tt::ConformsTo<Parallel::protocols::InitializationAction>`. Also add to their tests the line:
    ```cpp
    static_assert(tt::assert_conforms_to<TheAction, Parallel::protocols::InitializationAction>)`
    ```

Now the protocol defines what an "initialization action" is, provides documentation and gives an example.
Only actions that indicate they are initialization actions by conforming to the protocol are picked up by `SetupDataBox`. This establishes the strong link from the action to the documentation of what it means to be an initialization action, and makes it unit-testable that the action will work correctly with `SetupDataBox`. Let me know what you think :)

Reply:
I think that type of constraint can be useful, but some (particularly @nilsdeppe) have expressed reservations about extensive usage of the protocols because of challenges that can arise during debugging. Because this constraint would affect every initialization action (which can already be difficult to debug for other reasons), I'd like another voice in favor of using the protocols for this before implementing the change everywhere.
If we don't apply this immediately for this PR, we can open an issue indicating that this would be a good place to consider for protocols.

Reply:
I haven't tested the newer protocol revamps extensively, partly because I've been avoiding using protocols after my last experience. I would be mildly against making this switch because (a) I'm not yet completely convinced protocols work well enough and (b) it's a lot of work for something we don't know if it works well enough. I don't disagree that having some checks like what protocols offers (and the associated documentation) would be useful, but I'm hesitant to suggest that a +1600/-1000 line PR is the place to test this... I will say that you should _not_ put in in a `Protocols.hpp`. Having a single file to collect all protocols will not scale. We've tried that with tags and it didn't work.

Reply:
That's fair. I think this would be a very fitting use case for a protocol, but if you don't want to go that route for the reasons @nilsdeppe gave that's totally fine with me.

Reply:
I've made issue #2570 for now to list protocols to add once protocols are more thoroughly tested elsewhere.

Reply:
Sounds good. I added a very limited use of protocols to #2569 where it makes a lot of sense as well but doesn't affect such a large part of our codebase :)